File: tests/test_v7_execution_mode.cpp
#include "pm/v7_execution_mode.hpp"

#include <cassert>

int main() {
    using pm::v7::ExecutionMode;
    using pm::v7::execution_capabilities;
    using pm::v7::execution_mode_name;
    using pm::v7::legacy_execution_flags_match;
    using pm::v7::parse_execution_mode;

    const auto paper = parse_execution_mode("PAPER_SIMULATED");
    assert(paper && *paper == ExecutionMode::PaperSimulated);
    assert(execution_mode_name(*paper) == "PAPER_SIMULATED");
    const auto paper_capabilities = execution_capabilities(*paper);
    assert(paper_capabilities.public_ingress && paper_capabilities.paper_orders);
    assert(!paper_capabilities.authenticated_read && !paper_capabilities.signing);
    assert(!paper_capabilities.submit_orders && !paper_capabilities.cancel_orders);
    assert(legacy_execution_flags_match(*paper, true, false, false));
    assert(!legacy_execution_flags_match(*paper, false, false, false));

    const auto shadow = parse_execution_mode("SHADOW_LIVE_READ_ONLY");
    assert(shadow && execution_capabilities(*shadow).authenticated_read);
    assert(!execution_capabilities(*shadow).signing);
    assert(legacy_execution_flags_match(*shadow, false, true, false));

    const auto cancel = parse_execution_mode("CANCEL_ONLY");
    assert(cancel && execution_capabilities(*cancel).cancel_orders);
    assert(!execution_capabilities(*cancel).submit_orders);
    assert(!parse_execution_mode("PAPER_OR_LIVE"));
    return 0;
}

File: tests/test_v7_execution_mode.py
#!/usr/bin/env python3
import json
import unittest
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
MODES = {
    "RESEARCH_ZERO_AUTHORITY", "PAPER_SIMULATED", "SHADOW_LIVE_READ_ONLY", "MICRO_LIVE",
    "LIVE_RESTRICTED", "LIVE_SCALED", "DRAIN_ONLY", "CANCEL_ONLY", "KILLED",
}


class ExecutionModeContractTests(unittest.TestCase):
    def test_checked_in_execution_modes_are_explicit_and_zero_cap(self) -> None:
        value = json.loads((ROOT / "config/v7_execution_modes.json").read_text(encoding="utf-8"))
        self.assertEqual(value["schema"], "polymarket_v7_execution_modes_v1")
        self.assertEqual(value["default_execution_mode"], "PAPER_SIMULATED")
        self.assertEqual(set(value["modes"]), MODES)
        self.assertTrue(all(cap == 0 for cap in value["checked_in_live_caps"].values()))
        self.assertFalse(value["modes"]["PAPER_SIMULATED"]["authenticated_read"])
        self.assertFalse(value["modes"]["PAPER_SIMULATED"]["submit_orders"])

    def test_canonical_configs_agree_on_paper_simulated(self) -> None:
        paper = json.loads((ROOT / "config/paper_v7.json").read_text(encoding="utf-8"))
        supervision = json.loads((ROOT / "config/v7_runtime_supervision.json").read_text(encoding="utf-8"))
        champion = json.loads((ROOT / "config/live_champion.json").read_text(encoding="utf-8"))
        self.assertEqual(paper["execution_mode"], "PAPER_SIMULATED")
        self.assertEqual(paper["v7"]["execution_mode"], "PAPER_SIMULATED")
        self.assertEqual(paper["v7"]["execution_modes_policy"], "config/v7_execution_modes.json")
        self.assertEqual(supervision["execution_mode"], "PAPER_SIMULATED")
        self.assertEqual(champion["execution_mode"], "PAPER_SIMULATED")


if __name__ == "__main__":
    unittest.main()
